File: python/dry.py
# DRY (Don't Repeat Yourself)
# If a piece of code is identical or very similar to another, try to extract it into a generalized function

# Before
def greet_morning():
    print("Good morning, Alice!")

def greet_afternoon():
    print("Good afternoon, Alice!")

def greet_evening():
    print("Good evening, Alice!")

greet_morning()
greet_afternoon()
greet_evening()


# After
def greet(day_period, name):
    print(f"Good {day_period}, {name}!")

greet("morning", "Alice")
greet("afternoon", "Alice")
greet("evening", "Alice")

File: go/dry.go
// DRY (Don't Repeat Yourself)
// If a piece of code is identical or very similar to another, try to extract it into a generalized function

package main

import "fmt"

// Before
func greetMorning() {
	fmt.Println("Good morning, Alice!")
}

func greetAfternoon() {
	fmt.Println("Good afternoon, Alice!")
}

func greetEvening() {
	fmt.Println("Good evening, Alice!")
}

// After
func greet(dayPeriod, name string) {
	fmt.Printf("Good %s, %s!\n", dayPeriod, name)
}

func main() {
	// Before usage
	greetMorning()
	greetAfternoon()
	greetEvening()

	fmt.Println()

	// After usage
	greet("morning", "Alice")
	greet("afternoon", "Alice")
	greet("evening", "Alice")
}
